### PR TITLE
autoupdater: use blang/semver

### PR DESCRIPTION
**Pull request by mniewrzal**
What: This change updates auto-updater to use https://github.com/blang/semver as a tool to process version strings. 

Why: We want to be compliant with https://semver.org/

Please describe the tests:
 - Test 1: basic automatic test added
 
Please describe the performance impact: none

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: cmd/auto-updater/main_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+	tmpExec := ctx.File("storagenode")
+	err = ioutil.WriteFile(tmpExec, content, 0755)
+	require.NoError(t, err)
+
+	var mux http.ServeMux
+	content, err = ioutil.ReadFile("testdata/fake-storagenode.zip")
+	require.NoError(t, err)
+	mux.HandleFunc("/download", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write(content)
```
Comment (navillasa) [inline cmd/auto-updater/main_test.go]: maybe redundant comment but looks like jenkins linter is unhappy that this doesn't have an error check

Comment (mniewrzal) [reply cmd/auto-updater/main_test.go]: thanks, I missed that one

**Comment by mniewrzal** [comment]
> I guess I'm just curious what the main differences are between this new package and the previous `version.SemVer` values and methods used

In general our current support for version parsing is simple and in fact, we are not fully following  semantic versioning. This package implements parsing and processing (e.g. comparing) versions in a more comprehensive way. A decision about using it was made during the last arch review meeting.

File: cmd/auto-updater/main.go [hunk 2]
```diff
@@ -167,37 +162,50 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
...
 		if err := update(ctx); err != nil {
 			// don't finish loop in case of error just wait for another execution
 			log.Println(err)
 		}
 		return nil
-	})
+	}
+
+	if loopInterval == 0 {
+		err = loopFunc(ctx)
```
Comment (kaloyan-raev) [inline cmd/auto-updater/main.go]: Why do we have this special case? Does it mean that if `loopInterval == 0` then the check or update will be executed only once and the program will exit?

File: cmd/auto-updater/main.go [hunk 3]
```diff
@@ -167,37 +162,50 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
...
 	}
 
 	scanner := bufio.NewScanner(bytes.NewReader(out))
 	for scanner.Scan() {
 		line := scanner.Text()
 		prefix := "Version: "
 		if strings.HasPrefix(line, prefix) {
-			return version.NewSemVer(line[len(prefix):])
+			// use version string without 'v'
+			return semver.Make(line[len(prefix)+1:])
```
Comment (kaloyan-raev) [inline cmd/auto-updater/main.go]: How are we sure that in this case the version starts with `v`? Should we have an additional `strings.HasPrefix(line, "v")` check?

File: cmd/auto-updater/main.go (hunk 2 above)
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: yes, with loopInterval == 0 program will execute only once. Its a case for automatic test.

File: cmd/auto-updater/main.go (hunk 3 above)
Comment (mniewrzal) [reply cmd/auto-updater/main.go]: yeah, additional check makes sense. I will add it.

File: cmd/auto-updater/main.go (hunk 2 above)
Comment (kaloyan-raev) [reply cmd/auto-updater/main.go]: This sounds like an easter egg.

To me, loopInterval == 0 sound like a valid case for the Cycle to not wait between loop runs.

Can we check for a negative interval instead? If interval < 0 then run only once. And document it in the flag description.

Comment (mniewrzal) [reply cmd/auto-updater/main.go]: done

Comment (kaloyan-raev) [reply cmd/auto-updater/main.go]: Oh, shoot! I thought `0` is a valid value for `sync2.NewCycle`, but it is the `time.Ticker` that does not accept it. Then maybe check for `loopInterval <= 0` instead. Sorry about that!

File: cmd/auto-updater/main.go [hunk 4]
```diff
@@ -60,7 +61,7 @@ var (
 func init() {
 	rootCmd.AddCommand(runCmd)
 
-	runCmd.Flags().StringVar(&interval, "interval", "06h", "interval for checking the new version")
+	runCmd.Flags().StringVar(&interval, "interval", "06h", "interval for checking the new version, negative value will execute version check only once")
```
Comment (kaloyan-raev) [inline cmd/auto-updater/main.go]: This should now say "non-positive value" or "0 or less".

Comment (mniewrzal) [reply cmd/auto-updater/main.go]: ah yes :)

**Pull request merged**